src/librustc/traits/error_reporting.rs

```diff
@@ -663,13 +663,63 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                 if actual_trait_ref.self_ty().references_error() {
                     return;
                 }
-                struct_span_err!(self.tcx.sess, span, E0281,
-                    "type mismatch: the type `{}` implements the trait `{}`, \
-                     but the trait `{}` is required ({})",
-                    expected_trait_ref.self_ty(),
-                    expected_trait_ref,
-                    actual_trait_ref,
-                    e)
+                let expected_trait_ty = expected_trait_ref.self_ty();
+                if expected_trait_ty.is_closure() {
+                    if let &TypeError::TupleSize(ref expected_found) = e {
```

Comment:
Shouldn't you match on `expected_trait_ref` & `actual_trait_ref`, to avoid `Fn(X<()>)` vs. `Fn(X<(Foo, Bar)>)` etc.